### PR TITLE
perf(b-link): convert functional component to regular component (closes #3634)

### LINKED ISSUE #3634: b-dropdown performance problem since RC12
### Describe the bug

I have a b-table with 27 rows. Within each row I have a b-dropdown, which contains 18 b-dropdown-items (486 items in total). I have a completely independent b-form-group containing a b-form-input on the vue page. Note that the table is not relevant, the issue persists if I simply create 27 b-dropdowns with 18 items in each.

Since rc12, when I type (keydown) in the input box, the input is very laggy (the text only in appears in the box with significant latency after hitting the keys).

If I remove or significantly reduce the number of dropdown-items on the page, this issue goes away. On rc11 the input box performs perfectly with no issues.

### Steps to reproduce the bug

1. Create a page using my example, which I will paste at the bottom of this ticket
2. Open/load page
3. Type into the input box, note the lag
4. Revert to 2.0.0rc11, observe that there is no lag

### Expected behavior

The number of drop-down items on the page should not effect the keydown performance of the input box

### Versions

**Libraries:**

- BootstrapVue: 2.0.0.rc12
- Vue: 2.5.17

**Environment:**

- Device: Windows laptop
- OS: Windows
- Browser: Chrome
- Version: 75

### Additional context

Example vue page to recreate problem:

[dropdownPerformance.txt](https://github.com/bootstrap-vue/bootstrap-vue/files/3367916/dropdownPerformance.txt)

### PR DESCRIPTION
**Pull request by tmorehouse**
### Describe the PR

Convert `b-link` to a regular component from a functional component.

When users have a large number of components, which are derived from `b-link` (`b-dropdown-item`, `b-button`, etc), re-rendering can take performance hits due to the re-computation of the event handlers on each render.

Closes #3634

To do:
- [x] updated tests

### PR checklist

**What kind of change does this PR introduce?** (check at least one)

- [x] Bugfix
- [ ] Feature
- [ ] Enhancement
- [ ] ARIA accessibility
- [ ] Documentation update
- [x] Other (please describe): performance

**Does this PR introduce a breaking change?** (check one)

- [x] No
- [ ] Yes (please describe)

**The PR fulfills these requirements:**

- [x] It's submitted to the `dev` branch, **not** the `master` branch
- [x] When resolving a specific issue, it's referenced in the PR's title (i.e. `[...] (fixes #xxx[,#xxx])`, where "xxx" is the issue number)
- [x] It should address only one issue or feature. If adding multiple features or fixing a bug and adding a new feature, break them into separate PRs if at all possible.
- [x] The title should follow the [**Conventional Commits**](https://www.conventionalcommits.org/) naming convention (i.e. `fix(alert): not alerting during SSR render`, `docs(badge): update pill examples, fix typos`, `chore: fix typo in README`, etc). **This is very important, as the `CHANGELOG` is generated from these messages.**

**If new features/enhancement/fixes are added or changed:**

- [ ] Includes documentation updates (including updating the component's `package.json` for slot and event changes)
- [x] New/updated tests are included and passing (if required)
- [x] Existing test suites are passing
- [x] The changes have not impacted the functionality of other components or directives
- [ ] ARIA Accessibility has been taken into consideration (Does it affect screen reader users or keyboard only users? Clickable items should be in the tab index, etc.)

**If adding a new feature, or changing the functionality of an existing feature, the PR's
description above includes:**

- [ ] A convincing reason for adding this feature (to avoid wasting your time, it's best to open a suggestion issue first and wait for approval before working on it)

### CODE REVIEW

File: src/components/link/link.js [hunk 1]
```diff
@@ -112,78 +112,109 @@ export const omitLinkProps = propsToOmit => {
   }, {})
 }
 
-const clickHandlerFactory = ({ disabled, tag, href, suppliedHandler, parent }) => {
```
Comment (jackmu95) [inline src/components/link/link.js]: @tmorehouse I think we can remove the `pickLinkProps()` and `omitLinkProps()` utils and their tests as they aren't used anywhere in the codebase.

Comment (tmorehouse) [reply src/components/link/link.js]: Possibly... but I though they were used somewhere....

File: src/components/link/link.js [hunk 2]
```diff
@@ -112,78 +112,109 @@ export const omitLinkProps = propsToOmit => {
...
+        this.$root.$emit('clicked::link', evt)
+      }
+      if ((!isRouterLink && this.computedHref === '#') || this.disabled) {
+        // Stop scroll-to-top behavior or navigation on
+        // regular links when href is just '#'
+        evt && evt.preventDefault()
+      }
+    },
+    focus() {
+      this.$el && this.$el.focus && this.$el.focus()
```
Comment (jackmu95) [inline src/components/link/link.js]: I would prefer the following syntax:
```
if (this.$el && this.$el.focus) {
  this.$el.focus()
}
```

File: src/components/link/link.js [hunk 3]
```diff
@@ -112,78 +112,109 @@ export const omitLinkProps = propsToOmit => {
...
+    },
+    blur() {
+      this.$el && this.$el.blur && this.$el.blur()
```
Comment (jackmu95) [inline src/components/link/link.js]: Same as above

File: src/components/link/link.js (hunk 1 above)
Comment (jackmu95) [reply src/components/link/link.js]: I thought so too in the first place but I looked it up and haven't found any reference.

Comment (tmorehouse) [reply src/components/link/link.js]: They may have been used previously... and then removed.

And we have the newer `omit` helper in `utils/object.js` as well.

Comment (jackmu95) [reply src/components/link/link.js]: Yeah. Will save some bytes too...

Comment (tmorehouse) [reply src/components/link/link.js]: Removed the two helper utils... lets see if anything complains.

**Pull request merged**